partitura/musicanalysis/note_array_to_score.py

```diff
@@ -0,0 +1,358 @@
+
+    if not all(x in dtypes for x in ['step', 'alter', 'octave']):
+        warnings.warn("pitch spelling")
+        spelling_global = analysis.estimate_spelling(note_array)
+        note_array = rfn.merge_arrays((note_array, spelling_global), flatten=True)
+
+    if estimate_voice_info:
+        warnings.warn("voice estimation", stacklevel=2)
+        # TODO: deal with zero duration notes in note_array.
+        # Zero duration notes are currently deleted
+        estimated_voices = analysis.estimate_voices(note_array)
+        assert len(part_voice_list) == len(estimated_voices)
+        for i, (part_voice, voice_est) in enumerate(zip(part_voice_list, estimated_voices)):
+            # Not sure if correct.
+            if part_voice != np.inf:
+                estimated_voices[i] = part_voice
+        note_array = rfn.append_fields(note_array, "voice", np.array(estimated_voices, dtype=int))
+
+    if estimate_key or ('ks_fifths' not in dtypes and 'ks_mode' not in dtypes):
+        warnings.warn("key estimation", stacklevel=2)
+        k_name = analysis.estimate_key(note_array)
+        global_key_sigs = [[0, k_name]]
+    else:
+        if key_sigs is None and "ts_beats" in dtypes:
+            global_key_sigs = [[0, fifths_mode_to_key_name(note_array[0]["ks_fifths"], note_array[0]["ks_mode"])]]
+            for n in note_array:
+                if n["ts_beats"] != global_key_sigs[-1][1] or n["ts_beat_type"] != global_key_sigs[-1][2]:
+                    global_key_sigs.append([n["onset_div"], fifths_mode_to_key_name(n["ks_fifths"], n["ks_mode"])])
+            else:
+                global_key_sigs = key_sigs
+    global_key_sigs = np.array(global_key_sigs)
+    # for convenience, we add the end times for each time signature
+    ks_end_times = np.r_[global_key_sigs[1:, 0], np.max(note_array["onset_div"]+note_array["duration_div"])]
+    global_key_sigs = np.column_stack((global_key_sigs, ks_end_times))
+
+    # compute quarter divs
+    if divs is None:
+        # find first note with nonzero duration (in case score starts with grace_note).
+        for idx, dur in enumerate(note_array["duration_beat"]):
+            if dur != 0:
+                break
+        divs = int((note_array[idx]["duration_div"] / note_array[idx]["duration_beat"])*(note_array[idx]["ts_beat_type"]/4))
+
+    # Steps for dealing with anacrusis measure.
+    if np.all(anacrusis_mask == False):
+        anacrusis_divs = 0
+    else:
+        last_neg_beat = np.max(note_array[anacrusis_mask]["onset_beat"])
+        last_neg_divs = np.max(note_array[anacrusis_mask]["onset_div"])
+        beat_type = np.max(note_array[anacrusis_mask]["ts_beat_type"])
+        difference_from_zero = (0 - last_neg_beat) * divs * (4 / beat_type)
+        anacrusis_divs = int(last_neg_divs + difference_from_zero)
+
+    # Test again for negative divs
+    assert np.all(note_array["onset_div"] >= 0), "Negative divs found in note_array."
```

Comment:
why is a second test required?

Reply:
Second test is to verify no negative onset divs are present, which means that the input was faulty

Reply:
should it then go on top with the other input checks?